packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md

```diff
@@ -0,0 +1,111 @@
+
+Some of the cell state that would incorrectly apply to new rows (once the user scrolls) is:
+
+-   Selected text (which can span multiple rows), for both editable and non-editable cells
+-   Text cursor position (for editable controls in table cells)
+-   Pending input (for editable controls in table cells)
+-   Open popups (action menu or select/combobox cells whose popup is open)
+
+## Links To Relevant Work Items and Reference Material
+
+[#997: Table virtualization: Clear/Commit (and potentially save/restore) row/cell state when scrolling](https://github.com/ni/nimble/issues/997)
+
+## Implementation / Design
+
+When the user scrolls the table, we will:
+
+-   'Blur' (Lose focus) from any active/focused control in a table cell. Generally this means that whatever text the user began to change in that control will be committed, not discarded.
+-   If a cell action menu is open (in which case a menu item is focused), we'll close the associated menu via the menu button.
+
+We will not:
+
+-   Detect and deselect selected table text on scroll. This means that if table text is selected and the user scrolls, text will still be selected (potentially different text). We don't consider that to be a primary table interaction, and it shouldn't harm functionality (it's easy to resolve for the user by clicking off to deselect text). It's also the same behavior as the current `sl-grid`/`smart-table` used in SystemLink Enterprise.
+    -   See the "Alternative Implementations" section for additional information on this decision.
+-   Try and re-apply any state to the re-bound rows/cells after the scroll. That means that we won't re-focus the previously focused control in a cell / re-open an action menu, after a scroll operation.
+
+See [the prototype branch](https://github.com/ni/nimble/commit/57a85347f86aa55c416671f135c484ee96701416) and the [prototype table Storybook](https://60e89457a987cf003efc0a5b-jpkalylfjo.chromatic.com/iframe.html?args=data:LargeDataSet&id=table--table&viewMode=story) to illustrate the concepts discussed in the following sections.
+
+### Blur Focused Controls in Cells
+
+In this case, `document.activeElement` will be the Nimble `Table`, and `table.shadowRoot.activeElement` will be a Nimble `TableRow`. We can recursively look at the active element's `shadowRoot.activeElement`, starting from the TableRow, and stop when we reach a Nimble `TableCell` or `null` (see [prototype](https://github.com/ni/nimble/commit/57a85347f86aa55c416671f135c484ee96701416#diff-e2bfd4eda0a3c89f54b4e09624e6cfd5a68ea2f14c0e79f717eacce38ec5982bR134)).
+
+If we found a focused `TableCell`, we then want to `blur()` the control in it:  
+(**Open Question: Pick one of the following two approaches**)  
+**Option 1:**  
```

Comment:
My vote is for this option. I far prefer to have an API semantically oriented around the notion of `blur`, even if what we have to document is that a column implementer must manually call `blur` for the proper behavior.

Avoiding semantics associated with cell recycling in our public API seems highly desirable.

Reply:
+1 for blur semantics. +1 for template events for responding to table actions in column apis.

The table column api could be called onCellBlur instead of onBeforeFocusedCellRecycled in which case both the template event based approach and column callback approach have blur semantics.

But I prefer the template event based approach either way. It nicely encapsulates behavior in the template and keeps the column author api consistent as pretty dumb configuration objects.

Reply:
Current proposal is a new `onBeforeBlur()` method on the new `BaseCellElement` class (which is subclassed for each different column type). The state/behavior will all be in the new cell element, not in the column object.